File: backend_server/controllers/authController.js
const jwt = require('jsonwebtoken')
const { User } = require('../models/userModel')
const CustomError = require('../utils/CustomError')
const nodemailer = require('nodemailer')
const dotenv = require('dotenv')

dotenv.config()



const getUserFromToken = async (req, res, next) => {
    const authToken = await req.headers.authorization.split(' ')[1]
    if(!authToken) {
        return next(new CustomError('Auth token not found.', 404))
    }
    const decoded = jwt.verify(authToken, process.env.JWT_SECRET)

    const user = await User.findOne({ _id : decoded.id })
    if(!user){
        return next(new CustomError('User not found.', 404))
    }
    return user
}

exports.signUp = async (req, res, next) => {
    try {
        const newUser = await User.create({
            email: req.body.email,
            password: req.body.password
        })

        const token =  jwt.sign({ id : newUser._id}, process.env.JWT_SECRET)

        async function sendMail(){
            const transporter = nodemailer.createTransport({
                host: 'smtp.mailgun.org',
                port : 587,
                secure: false,
                auth : {
                    user : process.env.MAILGUN_USER,
                    pass : process.env.MAILGUN_PASSWORD
                }
            })
    
            let info = await transporter.sendMail({
                from: 'sushant@studyease.tech',
                to: newUser.email,
                subject: 'Welcome to Study Ease 😀',
                // text: `Here is your One Time Password : ${newUser.OTP}`,
                html: `
                <p>Hi,</p>

                <p>I'm Sushant, the developer of Study Ease. Thank you for creating an account in my app. I hope that my app will be beneficial to your studies.</p>
                
                <p>Here is your One-Time-Password:&#160; <b>${newUser.OTP}</b></p>
                
                <p>For any doubts or support, please email us at support@studyease.com.</p>
                
                <p>Join on WhatsApp: - <a href="https://chat.whatsapp.com/JA6TAD3foPnFKNIelGlS7z">https://chat.whatsapp.com/JA6TAD3foPnFKNIelGlS7z</a></p>
                
                <p>Thank You</p>`
            })
        }
    
        sendMail().catch(err => {
            return next(err)
        })

        return res.status(200).json({
            status: 'success',
            message: 'Your account has been created successfully.',
            token: token,
        })
    } catch (err) {
        return next(err)
    }
}

exports.OTPVerify = async (req, res, next) => {
    try {
        const OTP = req.body.OTP
        const user = await getUserFromToken(req, res, next)
        req.user = user

        if(!OTP){
            return next(new CustomError('Please enter the OTP.', 401))
        }
        const isOTPValid = await req.user.verifyOTP(OTP, req.user.OTP)

        if(isOTPValid) {
            await req.user.updateOne({accountStatus: 'pendingCompletion'})
            await req.user.updateOne({OTP : undefined})
            return res.status(200).json({
                status: 'success',
                message: 'Your account has been activated.'
            })
        } else {
            return res.status(401).json({
                status: 'fail',
                message: 'Invalid OTP'
            })
        }
        
    } catch (err) {
        return next(err)
    }
}

exports.completeProfile = async (req, res, next) => {
    const { name, phoneNumber, college } = req.body

    if(!phoneNumber || !college) {
        return next(new CustomError('Please provide required information.', 404))
    }

    const user = await getUserFromToken(req, res, next)
    req.user = user

    try {
        await req.user.updateOne({ name: name, phoneNumber: phoneNumber, college: college, accountStatus: 'active'})
    } catch(err) {
        return next(err)
    }

    return res.status(200).json({
        status: 'success',
        message: 'Profile completed successfully.'
    })
}

exports.login = async (req, res, next) => {
    const { email, password } = req.body

    if(!email || !password) {
        return next(new CustomError('Please provide a valid email and password.'), 400)
    }
    let user

    try {
        user = await User.findOne({ email }).select('+password')
        
        if(!user || !await user.checkPassword(password)) {
            return next(new CustomError('Entered email or password is incorrect.', 401))
        }
    } catch (err) {
        return next(err)
    }

    const token = jwt.sign({ id : user._id}, process.env.JWT_SECRET)

    req.user = user

    return res.status(200).json({
        status:'success',
        token: token,
        message: 'You have successfully signed in.',
        account: user
    })
}

exports.autoLogin = async (req, res, next) => {
    try {
        const user = await getUserFromToken(req, res, next)
        req.user = user

        return res.status(200).json({
            status: 'success',
            token: await req.headers.authorization.split(' ')[1],
            message: 'You have successfully signed in.',
            account: user
        })
    } catch (err) {
        return next(err)
    }
}

exports.forgotPassword = async (req, res, next) => {
    try {
        const { email } = req.body

        if(!email) {
            return next(new CustomError('Email not specified.', 401))
        }

        const user = await User.findOne({ email })

        if(!user){
            return next(new CustomError('No user found with given email.', 404))
        }

        if(user['accountStatus'] == 'disabled'){
            return next(new CustomError('Account disabled.', 400))
        }  

        try{
            await user.createPasswordResetOTP()
            user.save({validateBeforeSave : false})

            req.email = user.email

                    async function sendMail(){
                        const transporter = nodemailer.createTransport({
                            host: 'smtp.mailgun.org',
                            port : 587,
                            secure: false,
                            auth : {
                                user : process.env.MAILGUN_USER,
                                pass : process.env.MAILGUN_PASSWORD
                            }
                        })
    
                        let info = await transporter.sendMail({
                            from: 'sushant@studyease.tech',
                            to: user.email,
                            subject: 'Reset your Study Ease Passsword.',
                            html: `
                            <p>Hi,</p>
                            
                            <p>Here is your One-Time-Password:&#160; <b>${user.OTP}</b></p>
                            
                            <p>For any doubts or support, please email us at support@studyease.tech.</p>
                            
                            <p>Join on WhatsApp: - <a href="https://chat.whatsapp.com/JA6TAD3foPnFKNIelGlS7z">https://chat.whatsapp.com/JA6TAD3foPnFKNIelGlS7z</a></p>
                            
                            <p>Thank You</p>`  
                        })
                    }
    
                    sendMail().catch(err => {return next(err)})

            return res.status(200).json({
                status: 'success',
                message: 'An OTP has been created and sent to your email address.'
            })

        } catch(err){
            return next(err)
        }   

    } catch (err) {
        return next(err)
    }
}


exports.resetPassword = async (req, res, next) => {
    try {
        const OTP = req.body.OTP
        const newPassword = req.body.newPassword
        if(!OTP){
            return next(new CustomError('OTP not found.', 404))
        }

        const user = await User.findOne({ OTP })   

        if(!user){
            return next(new CustomError('Entered OTP is not valid.', 400))
        }

        try {
            user.password = newPassword

            await user.save().then(async () => {
                await user.resetOTP()
                user.save({ validateBeforeSave: true })
            })

            return res.status(200).json({
                status: 'success',
                message: 'Your password has been updated.'
            })

        } catch (err) {
            return next(err)
        }
        
    } catch (err) {
        return next(err)
    }
}

exports.protectAuth = async (req, res, next) => {
    try {
        let decoded
        let token

        if(req.headers.authorization && req.headers.authorization.startsWith('Bearer')){
            token = await req.headers.authorization.split(' ')[1]
        }
        if(token === undefined){
            return next(new CustomError('You are not logged in.', 401))
        }
        try {
            decoded = jwt.verify(token, process.env.JWT_SECRET)
        } catch (err) {
            return next(err)
        }

        const freshUser = await User.findOne({id: decoded._id})
        if(!freshUser) return next(new CustomError('User has been deleted.', 401))

        req.user = freshUser
        next()
    } catch (err) {
        return next(err)
    }
}


exports.protectAdmin = async (req, res, next) => {
    try {
        let token
        if(req.headers.authorization && req.headers.authorization.startsWith('Bearer')){
            token = await req.headers.authorization.split(' ')[1]
        }
        if(token === undefined){
            return next(new CustomError('You are not logged in.', 401))
        }

        const decodedToken = jwt.verify(token, process.env.JWT_SECRET)
        const user = await User.findById(decodedToken.id)
        
        if(user.isAdmin != true) return next(new Error('Action not allowed.', 405))

        next()
    } catch (err) {
        return next(err)
    }
}

File: study_ease/lib/controllers/auth/auth_controller.dart
import 'dart:async';
import 'dart:convert';
import 'package:get/get.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:study_ease/controllers/applovin/applovin_controller.dart';
import 'package:study_ease/controllers/dialogs/dialogs_controller.dart';
import 'package:study_ease/models/user/user_model.dart';
import 'package:restart_app/restart_app.dart';
import 'package:study_ease/views/auth/createAccount/create_account.dart';

Map<String, String> requestHeaders = {
  'Content-type': 'application/json',
  'Accept': 'application/json'
};

class AuthController extends GetxController {
  String usersAPIBaseURL = 'http://139.59.42.124:3000/api/v1/users';

  final dialogsController = Get.put(DialogsController());
  final applovinController = Get.put(ApplovinController());

  Future<dynamic> createAccount(String email, String password) async {
    if (email.isEmpty || !email.isEmail) {
      return dialogsController.showErrorDialog(
          'Incorrect Email', 'Please provide an email address.');
    }

    if (password.isEmpty || password.length < 8) {
      return dialogsController.showErrorDialog('Weak password',
          'Make sure that your password is at least 8 characters.');
    }

    final prefs = await SharedPreferences.getInstance();

    dialogsController.showLoadingDialog();

    var response = await http.post(Uri.parse('$usersAPIBaseURL/sign-up'),
        headers: requestHeaders,
        body: jsonEncode({'email': email, 'password': password}));

    dialogsController.hideDialog();

    // Checking rate limit
    if (response.statusCode == 429) {
      return dialogsController.showErrorDialog('Too many requests',
          'You are sending too many requests to the server. Please try again after 30 minutes.');
    }

    var data = jsonDecode(response.body);

    if (data['message'] == 'Your account has been created successfully.') {
      dialogsController.hideDialog();
      await prefs.setString('authToken', data['token']);
      Get.offAllNamed('otp');
    } else if (data['message'] == 'Duplicate Value.') {
      return dialogsController.showErrorDialog(
          'Duplicate E-mail', 'Email already exists. Please log in instead.');
    } else {
      return dialogsController.showErrorDialog();
    }
  }

  // ---------------------------------------------------------------- OTP Verify
  Future<dynamic> verifyOTP(String enteredOTP) async {
    if (enteredOTP.isEmpty) {
      return dialogsController.showErrorDialog(
          'Empty OTP', 'Make sure to provide an OTP.');
    }

    final prefs = await SharedPreferences.getInstance();
    final authToken = prefs.getString('authToken');

    Map<String, String> authorizedHeaders = {
      'Content-type': 'application/json',
      'Accept': 'application/json',
      'authorization': 'Bearer $authToken'
    };

    if (!enteredOTP.isNum) {
      return dialogsController.showErrorDialog(
          'Error', 'Please enter a numeric OTP only.');
    }
    var otp = int.parse(enteredOTP);

    dialogsController.showLoadingDialog();
    var response = await http.post(Uri.parse('$usersAPIBaseURL/verify-otp'),
        headers: authorizedHeaders, body: jsonEncode({'OTP': otp.toInt()}));
    dialogsController.hideDialog();

    var data = jsonDecode(response.body);

    if (data['message'] == 'Your account has been activated.') {
      Get.offAllNamed('completeProfile');
    } else if (data['message'] == 'Invalid OTP') {
      return dialogsController.showErrorDialog(
          data['message'], 'Entered OTP is invalid. Please type carefully.');
    } else if (data['message'] == 'Please enter the OTP.') {
      return dialogsController.showErrorDialog(
          data['message'], 'Entered OTP is invalid. Please type carefully.');
    } else {
      dialogsController.showErrorDialog();
    }
  }

  // ---------------------------------------------------------------- Complete Profile
  Future<dynamic> completeProfile(
      String name, String phoneNumber, String college) async {
    if (name.isEmpty || phoneNumber.isEmpty || college.isEmpty) {
      return dialogsController.showErrorDialog(
          'Error', 'Please make sure to provide correct details.');
    }

    final prefs = await SharedPreferences.getInstance();
    final authToken = prefs.getString('authToken');

    Map<String, String> authorizedHeaders = {
      'Content-type': 'application/json',
      'Accept': 'application/json',
      'authorization': 'Bearer $authToken'
    };

    if (!phoneNumber.isNum) {
      return dialogsController.showErrorDialog(
          'Error', 'Please enter a phone number.');
    }

    var phone = int.parse(phoneNumber);

    dialogsController.showLoadingDialog();
    var response = await http.post(
        Uri.parse('$usersAPIBaseURL/complete-profile'),
        headers: authorizedHeaders,
        body: jsonEncode(
            {"name": name, "phoneNumber": phone, "college": college}));
    dialogsController.hideDialog();

    var data = jsonDecode(response.body);
    var statusCode = response.statusCode;

    if (data['message'] == 'Profile completed successfully.') {
      autoLogin();
    } else if (statusCode != 200) {
      dialogsController.showErrorDialog();
    }
  }

  // ---------------------------------------------------------------- Login
  Future<dynamic> login(String email, String password) async {
    if (email.isEmpty || !email.isEmail) {
      return dialogsController.showErrorDialog(
          'Incorrect Email', 'Please provide an email address.');
    }

    if (password.isEmpty || password.length < 8) {
      return dialogsController.showErrorDialog(
          'Invalid Password', 'Please provide an valid password.');
    }

    final prefs = await SharedPreferences.getInstance();
    dialogsController.showLoadingDialog();

    var response = await http.post(Uri.parse('$usersAPIBaseURL/login'),
        headers: requestHeaders,
        body: jsonEncode({'email': email, 'password': password}));

    dialogsController.hideDialog();

    var data = jsonDecode(response.body);

    if (data['status'] == 'Not Found') {
      return dialogsController.showErrorDialog('Not Found',
          'Entered email & password is either wrong or there is no account with these details.');
    } else if (data['account']['accountStatus'] == 'disabled') {
      return dialogsController.showErrorDialog('Account Banned',
          'Your account has been disabled by the system. Contact us for more details.');
    } else if (data['status'] == 'success' &&
        data['message'] == 'You have successfully signed in.') {
      await prefs.setString('authToken', data['token']);
      await prefs.setString('userID', data['account']['id']);
      await prefs.setString('userName', data['account']['name']);
      await prefs.setString('userEmail', data['account']['email']);
      await prefs.setInt('userPhoneNumber', data['account']['phoneNumber']);
      await prefs.setString('userCollege', data['account']['college']);

      Get.offAllNamed('home');
    } else {
      dialogsController.showErrorDialog();
    }
  }

  // ---------------------------------------------------------------- Auto Login
  Future<dynamic> autoLogin() async {
    final prefs = await SharedPreferences.getInstance();
    final authToken = prefs.getString('authToken');

    Map<String, String> authorizedHeaders = {
      'Content-type': 'application/json',
      'Accept': 'application/json',
      'authorization': 'Bearer $authToken'
    };

    var response = await http.post(Uri.parse('$usersAPIBaseURL/auto-login'),
        headers: authorizedHeaders);

    var data = jsonDecode(response.body);

    if (data['message'] == 'User not found.') {
      return Get.offAllNamed('createAccount');
    } else if (data['account']['accountStatus'] == 'disabled') {
      return dialogsController.showErrorDialog(
          'Account Banned',
          'Your account has been disabled by the system. Contact us for more details.',
          () => {Get.offAllNamed('createAccount')});
    } else if (data['account']['accountStatus'] == 'pending') {
      Get.offAllNamed('otp');
    } else if (data['account']['accountStatus'] == 'pendingCompletion') {
      Get.offAllNamed('completeProfile');
    } else if (data['message'] == 'You have successfully signed in.') {
      await prefs.setString('authToken', data['token']);
      await prefs.setString('userID', data['account']['id']);
      await prefs.setString('userName', data['account']['name']);
      await prefs.setString('userEmail', data['account']['email']);
      await prefs.setInt('userPhoneNumber', data['account']['phoneNumber']);
      await prefs.setString('userCollege', data['account']['college']);
      Get.offAllNamed('home');
    } else {
      Get.offAllNamed('createAccount');
    }
  }

  // ---------------------------------------------------------------- Get User Details
  Future<User> getUserDetails() async {
    final prefs = await SharedPreferences.getInstance();
    final authToken = prefs.getString('authToken');

    Map<String, String> authorizedHeaders = {
      'Content-type': 'application/json',
      'Accept': 'application/json',
      'authorization': 'Bearer $authToken'
    };

    var response = await http.post(Uri.parse('$usersAPIBaseURL/auto-login'),
        headers: authorizedHeaders);

    return User.fromJson(jsonDecode(response.body));
  }

  //  ---------------------------------------------------------------- Forgot Password
  Future<dynamic> forgotPassword(String email) async {
    if (email.isEmpty || !email.isEmail) {
      return dialogsController.showErrorDialog(
          'Incorrect Email', 'Please provide an email address.');
    }

    dialogsController.showLoadingDialog();
    var response = await http.post(
        Uri.parse('$usersAPIBaseURL/forgot-password'),
        headers: requestHeaders,
        body: jsonEncode({'email': email}));
    dialogsController.hideDialog();

    // Checking rate limit
    if (response.statusCode == 429) {
      return dialogsController.showErrorDialog('Too many requests',
          'You are sending too many requests to the server. Please try again after 30 minutes.');
    }

    var data = jsonDecode(response.body);

    if (data['message'] ==
        'An OTP has been created and sent to your email address.') {
      Get.offAllNamed('resetPassword');
    } else if (data['message'] == 'No user found with given email.') {
      return dialogsController.showErrorDialog('Error',
          'No user found with given email. Make sure you entered correct email.');
    } else if (data['message'] == 'Account disabled.') {
      return dialogsController.showErrorDialog(
          'Account Disabled', 'Your account has been disabled.');
    } else {
      dialogsController.showErrorDialog();
    }
  }

  //  ---------------------------------------------------------------- Reset Password
  Future<dynamic> resetPassword(String enteredOTP, String newPassword) async {
    if (enteredOTP.isEmpty) {
      return dialogsController.showErrorDialog(
          'Error', 'OTP is empty. Make sure you enter an OTP.');
    }

    if (newPassword.isEmpty) {
      return dialogsController.showErrorDialog('Error',
          'New password is empty. Make sure you enter a new password.');
    }

    var otp = int.parse(enteredOTP);

    dialogsController.showLoadingDialog();
    var response = await http.post(Uri.parse('$usersAPIBaseURL/reset-password'),
        headers: requestHeaders,
        body: jsonEncode({'OTP': otp, 'newPassword': newPassword}));
    dialogsController.hideDialog();

    var data = jsonDecode(response.body);

    if (data['message'] == 'Entered OTP is not valid.') {
      return dialogsController.showErrorDialog('Error',
          'Entered OTP is not valid. Make sure to type OTP carefully.');
    } else if (data['message'] == 'Your password has been updated.') {
      Get.offAllNamed('login');
      Get.snackbar('Password Changed!', 'Your password has been updated.');
    } else {
      dialogsController.showErrorDialog();
    }
  }

  //////////////////////////////////////////////////////////////// Edit Profile

  Future<dynamic> editProfile(String newName, String newCollege) async {
    if (newName.isEmpty || newCollege.isEmpty) {
      return dialogsController.showErrorDialog(
          'Incorrect Data', 'Make sure to enter both name and college.');
    }

    final prefs = await SharedPreferences.getInstance();
    final authToken = prefs.getString('authToken');
    final userID = prefs.getString('userID');

    Map<String, String> authorizedHeaders = {
      'Content-type': 'application/json',
      'Accept': 'application/json',
      'authorization': 'Bearer $authToken'
    };

    dialogsController.showLoadingDialog();
    var response = await http.post(Uri.parse('$usersAPIBaseURL/edit-profile'),
        headers: authorizedHeaders,
        body: jsonEncode(
            {'newName': newName, 'newCollege': newCollege, 'userID': userID}));
    dialogsController.hideDialog();

    if (response.statusCode == 200) {
      Restart.restartApp();
    } else {
      return dialogsController.showErrorDialog();
    }
  }

  //////////////////////////////////////////////////////////////// LogOut

  Future<dynamic> logOut() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.clear().then((value) => {Get.offAll(const CreateAccount())});
  }
}
